feat: steering to avoid chain breaks and clashes

ludwigwinkler opened the pull request:
Add Steering functionality to BioEmu

ludwigwinkler committed:
feat: add TODO for fk steering implementation in heun_denoiser

ludwigwinkler committed:
Merge branch 'main' into luwinkler/fk_steering

ludwigwinkler committed:
gitignore test outputs and test case for steering

ludwigwinkler committed:
Merge branch 'luwinkler/fk_steering' of os.github.com:microsoft/bioemu into luwinkler/fk_steering

ludwigwinkler committed:
first prototype

ludwigwinkler committed:
feat: implement Kabsch alignment and enhance steering functionality with FK sampling

ludwigwinkler committed:
feat: add wandb to .gitignore to exclude Weights and Biases files

ludwigwinkler committed:
Refactor steering module and add new steering run script

- Enhanced the `steering.py` module with additional plotting functions for Ca-Ca distances and clashes.
- Introduced a new `steering_run.py` script for testing sample generation with steering, utilizing Hydra for configuration management.
- Created a scratch pad script for testing loss functions visually.
- Updated the test suite in `test_steering.py` to include WandB logging and improved configuration handling with Hydra.
- Removed deprecated code and organized potential classes for better clarity and maintainability.

ludwigwinkler committed:
Add BioEMU development rules and enhance load_md script

- Introduced a new `bioemu.mdc` file containing development guidelines for the BioEMU project, covering molecular dynamics, blob storage, curated MD data structure, file paths, analysis patterns, and error handling.
- Added a `load_md.py` script demonstrating how to interact with blob storage and load molecular dynamics data, including trajectory analysis using MDTraj.
- Updated the `run_steering_comparison.py` script to iterate over different particle counts for steering experiments, improving configurability and analysis capabilities.
- Enhanced the `denoiser.py` with a new `euler_maruyama_denoiser` function and integrated it into the testing framework.
- Updated configuration files for steering and denoising to reflect new parameters and functionalities.

ludwigwinkler committed:
Enhance BioEMU documentation and introduce new analytical diffusion module

- Updated the `bioemu.mdc` file to provide a comprehensive development guide, including architectural principles, design patterns, and implementation guidelines for the BioEMU project.
- Added a new `analytical_diffusion.py` module that implements a time-dependent Gaussian Mixture Model for analytical diffusion, including functionality for forward and reverse diffusion processes.
- Refactored the `load_md.py` script by removing unused imports to streamline the code.
- Enhanced the `run_steering_comparison.py` script to improve configurability and analysis of steering experiments, including adjustments to plotting and data handling.
- Introduced a new `stratified_sampling.py` module with tests for stratified resampling functionality.
- Added a `sweep_analysis.py` module for analyzing sweep data from Weights and Biases, including visualization of results.
- Updated the `steering.yaml` configuration to reflect changes in potential parameters for steering functionality.

Commit:
Update .gitignore to exclude fasta files in notebooks directory

Commit:
Update .gitignore to exclude output files from notebooks directory

Commit:
Refactor tqdm imports in analytical_diffusion, run_steering_comparison, and denoiser scripts

- Changed import statement for tqdm from `tqdm.auto` to `tqdm` for consistency across modules.
- Added plt.show() in analyze_termini_distribution function to ensure plots are displayed.
- Commented out plt.show() in main function to prevent automatic display during batch processing.

Commit:
Remove .cursor/ from tracking and add to .gitignore

ludwigwinkler committed:
Delete .cursor/rules directory

Commit:
Update .gitignore to exclude all files in .cursor directory

Commit:
Merge branch 'luwinkler/fk_steering' of github.com:microsoft/bioemu into luwinkler/fk_steering

Commit:
Enhance steering functionality and introduce new potentials

- Refactored steering module to include ChainBreakPotential and ChainClashPotential, replacing previous distance potentials.
- Updated run_steering_comparison.py to refine steering configurations, including adjustments to num_samples and particle counts.
- Implemented fast steering optimization to delay particle creation until steering start time for improved performance.
- Added validation for steering configurations and assertions to ensure expected steering execution.
- Introduced comprehensive tests for new steering features, including physical and fast steering capabilities.
- Updated steering.yaml configuration to reflect new potential parameters and added end time for steering.

Commit:
first workin prototype of fast_steering

Commit:
Enhance steering capabilities and update documentation

- Introduced a new section in the README for "Steering for Enhanced Physical Realism," detailing the use of Sequential Monte Carlo for guiding protein structure diffusion.
- Added example commands for enabling steering via CLI and Python API, including key parameters and potential configurations.
- Created a new `hydra_run.py` script for running BioEMU sampling with Hydra configuration management, allowing for easier experimentation with steering parameters.
- Updated existing scripts to reflect changes in steering configuration, including renaming parameters for clarity and consistency.
- Added a new `README_hydra_run.md` to document the usage of the Hydra-based entry point.
- Implemented tests for CLI integration, ensuring that steering functionality works as expected through command-line parameters.

Commit:
Refine steering documentation and enhance configuration handling

- Updated the README to clarify the steering process, including the default behavior for steering potentials and the use of multiple particles.
- Removed references to the now-optional `steering_potentials_config` parameter in example commands and clarified its default behavior.
- Enhanced the sample.py script to load default steering potentials when no custom configuration is provided, improving usability.
- Added warnings for missing default configuration files to aid in troubleshooting.

Commit:
Update README to refine steering section and enhance CLI usage instructions

- Changed the section title from "Steering for Enhanced Physical Realism" to "Steering structures" for clarity.
- Updated CLI instructions to specify the requirement of setting `--num_steering_particles` to greater than 1 for enabling steering.
- Removed Python API example for steering to streamline the documentation and focus on CLI usage.

Commit:
Update README to include Python API example for steering and remove Hydra configuration section

- Added a Python API example for steering, demonstrating how to use the `bioemu.sample` module.
- Removed the section detailing the Hydra configuration interface to streamline the documentation and focus on the primary usage methods.

Commit:
Update .gitignore to exclude documentation files

- Added an entry to .gitignore to ignore all files in the docs directory, preventing them from being tracked by Git.

Commit:
Refactor steering configuration and enhance disulfide bridge potential

- Updated the `run_steering_comparison.py` and `run_guidance_steering_comparison.py` scripts to streamline steering configuration handling.
- Introduced a new `DisulfideBridgePotential` class for guiding disulfide bridge formation, including parameters for specified cysteine pairs.
- Added a new configuration file for disulfide steering and updated existing steering configurations to reflect changes in potential definitions.
- Enhanced the `sample.py` module to support the new steering configuration structure, allowing for better integration of disulfide bridge steering.
- Implemented tests for the `DisulfideBridgePotential` to ensure correct functionality and energy calculations.

Commit:
Add guidance steering capabilities and update experiment scripts

- Introduced new guidance steering configuration and potential for enhanced structural constraints.
- Updated `run_guidance_steering_comparison.py` to support three-way comparison: no steering, resampling only, and guidance steering.
- Added new binary images for visualization of steering comparisons.
- Refactored `run_steering_experiment` to accommodate the new experiment type parameter.
- Enhanced analysis functions to compare termini distances and KL divergence across different steering methods.
- Updated existing steering configurations to include guidance steering options and parameters.
- Improved error handling and logging for better user feedback during experiments.

Commit:
Enhance guidance steering implementation and update experiment configurations

- Updated `run_guidance_steering_comparison.py` to streamline handling of guidance steering alongside resampling.
- Refactored steering configuration to improve clarity and functionality, including adjustments to learning rates and steps for guidance.
- Modified `dpm_solver` to apply gradient guidance more effectively and ensure correct score calculations.
- Increased sample size in main configuration for better statistical analysis during experiments.
- Updated binary image for visualization of steering comparisons.

Commit:
Enhance guidance steering comparison and update configurations

- Updated `disulfide_steering_example.py` to include comprehensive steering configurations and improved error handling for guidance steering.
- Added functionality to create and demonstrate various steering configurations, including no steering, resampling only, and guidance steering.
- Enhanced statistical analysis and visualization of Cα-Cα distances for disulfide bridge pairs across different steering methods.
- Introduced new binary images for output visualization of steering comparisons.
- Refactored `run_guidance_steering_comparison.py` to support dynamic parameter adjustments for guidance strength and particle counts during experiments.

YuuuXie committed:
Modify final resampling step in denoiser.py

Update final resampling logic to use real x0 instead of predicted x0 and print relevant message.

sarahnlewis commented on README.md:
```diff
@@ -65,6 +66,54 @@ By default, unphysical structures (steric clashes or chain discontinuities) will
+
+BioEmu includes a steering system that uses Sequential Monte Carlo (SMC) to guide the diffusion process toward more physically plausible protein structures. Steering applies potential energy functions during denoising to favor conformations that satisfy physical constraints. Algorithmically, steering simulates *multiple particles* per desired sample and resamples between these particles according ...
+
+### Quick Start with Steering
```
```suggestion
### Quick start with steering
```

sarahnlewis commented on README.md:
```diff
@@ -65,6 +66,54 @@ By default, unphysical structures (steric clashes or chain discontinuities) will
+- `resampling_freq`: How often to resample particles (default: 1)
+- `steering_potentials_config`: Path to potentials configuration file (optional, defaults to physical_potentials.yaml)
+
+### Available Potentials
```
```suggestion
### Available potentials
```

sarahnlewis commented on README.md:
```diff
@@ -65,6 +66,54 @@ By default, unphysical structures (steric clashes or chain discontinuities) will
 
+## Steering structures
+
+BioEmu includes a steering system that uses Sequential Monte Carlo (SMC) to guide the diffusion process toward more physically plausible protein structures. Steering applies potential energy functions during denoising to favor conformations that satisfy physical constraints. Algorithmically, steering simulates *multiple particles* per desired sample and resamples between these particles according ...
```
I suggest including a reference. Also, spell out that no guidance term is added to the state updates; we only do resampling.

YuuuXie committed:
fix x0 type

sarahnlewis reviewed (commented):
I started leaving some nitty comments but then realised review is maybe premature. Let's discuss 1:1.

Maybe it would make sense to split this into two PRs: 1. add hydra interface, 2. add steering.

I'd suggest returning all samples from each group of particles instead of following [boltz]( https://github.com/jwohlwend/boltz/blob/cb04aeccdd480fd4db707f0bbafde538397fa2ac/src/boltz/model/modules/diffusionv2.py#L486) and only returning 1 sample from each. It will be less wasteful and will make the logic around 'batch size' less confusing.  Also, because you use stratified resampling, there might be more useful diversity within each particle group than there is there.

Commit:
Add physicality steering comparison notebook and update configurations

- Introduced `physicality_steering_comparison.py` to facilitate experiments comparing steering methods on protein sequences.
- Implemented a new `run_steering_experiment` function to handle both steered and non-steered sampling.
- Enhanced the `termini_steering_comparison.py` notebook to analyze termini distance distributions and KL divergence between steered and non-steered samples.
- Updated configurations in `bioemu.yaml` and added a new `physical_steering.yaml` for steering parameters.
- Refactored existing code to improve clarity and maintainability, including adjustments to the handling of steering configurations and potential definitions.
- Added tests for new functionalities to ensure correctness and performance of the steering methods.

Commit:
Merge branch 'luwinkler/cli_steering' of github.com:microsoft/bioemu into luwinkler/cli_steering

ludwigwinkler committed:
Update README.md

Co-authored-by: Sarah Lewis <9419264+sarahnlewis@users.noreply.github.com>

ludwigwinkler committed:
Update README.md

Co-authored-by: Sarah Lewis <9419264+sarahnlewis@users.noreply.github.com>

ludwigwinkler committed:
Update README.md

Co-authored-by: Sarah Lewis <9419264+sarahnlewis@users.noreply.github.com>

Commit:
refs for readme on steering and smc

Commit:
Merge branch 'luwinkler/cli_steering' of github.com:microsoft/bioemu into luwinkler/cli_steering

ludwigwinkler committed:
Update notebooks/README_hydra_run.md

Co-authored-by: Sarah Lewis <9419264+sarahnlewis@users.noreply.github.com>

Commit:
some renaming files and typos

Commit:
Merge branch 'luwinkler/cli_steering' of github.com:microsoft/bioemu into luwinkler/cli_steering

Commit:
Remove unused guidance images: deleted `guidance_steering_comparison.png` and `guidance_visualization_display.png` to clean up the repository.

Commit:
Remove profiler module: deleted `src/bioemu/profiler.py` as it is no longer needed

Commit:
Remove unused steering scratch pad: deleted `src/steering_scratch_pad.py` as it is no longer needed.

Commit:
Remove unused denoiser configuration: deleted `src/bioemu/config/denoiser/em.yaml` as it is no longer needed.

Commit:
Remove unused guidance steering configuration: deleted `src/bioemu/config/steering/guidance_steering.yaml` as it is no longer needed.

ludwigwinkler committed:
Merge branch 'main' into luwinkler/cli_steering

YuuuXie commented on src/bioemu/config/denoiser/dpm.yaml:
```diff
@@ -2,5 +2,5 @@ _target_: bioemu.shortcuts.dpm_solver
-N: 50
-noise: 0.0
+N: 100
+noise: 0.5 # original dpm =0 for ode
```
Do you find when the noise is increased to 0.5 it does not work with 50 steps?

YuuuXie replied:
I'd suggest to keep the original setting (N=50, noise=0) as default. On the other hand, we can override the `denoiser` setting in `bioemu.yaml` if we use it as a steering example

YuuuXie commented on pyproject.toml:
```diff
@@ -22,6 +22,7 @@ dependencies = [
     "stackprinter",
     "typer",
     "uv",
+    "einops"
```
I don't see `einops` is used anywhere, do we need it?

YuuuXie commented on src/bioemu/convert_chemgraph.py:
```diff
@@ -2,12 +2,16 @@
 from pathlib import Path
+import os
 
+from matplotlib.pylab import f
```
remove unused imports

YuuuXie commented on src/bioemu/convert_chemgraph.py:
```diff
@@ -462,11 +702,17 @@ def save_pdb_and_xtc(
-        pos=pos_angstrom[0],
-        node_orientations=node_orientations[0],
+        pos=pos_angstrom[-1],
+        node_orientations=node_orientations[-1],
```
why does it save the last frame instead of the first? (I guess in terms of topology it does not matter, but just curious if anything does not work)

sarahnlewis replied:
+1

YuuuXie commented on src/bioemu/convert_chemgraph.py:
```diff
@@ -477,11 +723,23 @@ def save_pdb_and_xtc(
+        )
+        # Filtering ratio computed but not logged
+
         filtered_traj = filter_unphysical_traj(traj)
```
`filter_unphysical_traj` are called twice (line 732 and 743), and the logging/printing is duplicated

YuuuXie commented on src/bioemu/denoiser.py:
```diff
@@ -1,16 +1,37 @@
+import time
+import torch.autograd.profiler as profiler
+from torch.profiler import profile, ProfilerActivity, record_function
+from tqdm.notebook import tqdm
```
Had an impression this `tqdm.notebook` only works within the notebook, but maybe I'm wrong. It might be more general to just use `tqdm`

YuuuXie commented on src/bioemu/denoiser.py:
```diff
@@ -67,7 +88,7 @@ def update_given_drift_and_diffusion(
             mean = x + drift * dt
             sample = mean + self.noise_weight * diffusion * torch.sqrt(dt.abs()) * z
-        return sample, mean
+        return sample, mean, z
```
Do I understand correctly this `z` is not used anywhere, if so, we can just keep the original return without `z`?

YuuuXie commented on src/bioemu/steering.py:
```diff
@@ -0,0 +1,873 @@
+
+
+@torch.enable_grad()
+def potential_gradient_minimization(x, potentials, learning_rate=0.1, num_steps=20):
```
Shall we remove it if we don't have plan to include in this release?

YuuuXie reviewed (commented):
Thanks for the great efforts! I left some comments, and still need to take a closer look at the steering code, potential definitions etc. 

Is there a plan of splitting the current one into two PRs as Sarah suggested? Or this is already one of them?

sarahnlewis commented on README.md:
```diff
@@ -65,6 +66,54 @@ By default, unphysical structures (steric clashes or chain discontinuities) will
+- `num_steering_particles`: Number of particles per sample (1 = no steering, >1=steering)
+- `steering_start_time`: When to start steering (0.0-1.0, default: 0.0)
+- `steering_end_time`: When to stop steering (0.0-1.0, default: 1.0)
+- `resampling_freq`: How often to resample particles (default: 1)
```
Is this an interval or a frequency? Does resampling_freq=3 mean resampling after every 3 denoising steps?

sarahnlewis replied:
If resampling_freq=n means resampling every n denoising steps then this setting is _inverse_ of frequency, and is misnamed. You could call it resampling_interval instead.

sarahnlewis commented on README.md:
```diff
@@ -65,6 +66,54 @@ By default, unphysical structures (steric clashes or chain discontinuities) will
+
+### Available potentials
+
+When steering is enabled (num_steering_particles > 1) and no additional `steering_potentials_config.yaml` is provided, BioEMU automatically loads `physical_potentials.yaml` by default, which includes:
```
```suggestion
When steering is enabled (num_steering_particles > 1) and `steering_potentials_config` is not given, BioEmu automatically loads `physical_potentials.yaml` by default, which includes:
```
please `git grep BioEMU` and replace with `BioEmu`.

sarahnlewis commented on notebooks/disulfide_steering_example.py:
```diff
@@ -0,0 +1,253 @@
+        if key == "disulfide":
+            potentials_config[key]["guidance_steering"] = enable_guidance
+        else:
+            potentials_config[key]["guidance_steering"] = enable_guidance
```
The 'if' and 'else' blocks are identical.

sarahnlewis reviewed (commented):
Thanks for the new feature! I haven't finished reviewing yet but am leaving some comments on what I looked at so far.

Commit:
Update .gitignore to exclude additional files, enhance pyproject.toml with new dependencies, ensure output directory creation in convert_chemgraph.py, and modify import statements in sample.py for improved script execution.

Commit:
simplified steering logic and first prototype for integrated sampling and evaluation

Commit:
cleaned up notebooks and configs

Commit:
Refactor steering configuration and update README for clarity. Removed deprecated steering parameters from YAML files, streamlined steering logic in the codebase, and enhanced documentation to reflect changes in steering functionality and usage. Updated .gitignore to exclude additional files and improved performance of batch processing functions.

Commit:
Remove commented TODO regarding embedding file copying in get_embeds.py and clean up logging output for cached embeddings.

Commit:
fix formatting

Commit:
Fix save_pdb_and_xtc function to use the first element of pos_angstrom and node_orientations instead of the last, ensuring correct data is written to .pdb files.

Commit:
fixing small deviations in the code to reduce mental load of the reviewers

Commit:
fixing diverging code

Commit:
Update steering parameters in README.md to reflect new default values for start and end times.

sarahnlewis commented on .gitignore:
```diff
@@ -132,3 +134,22 @@ cython_debug/
+.cursor/
+.cursor/** */
+docs/*
+uv.lock
```
??

sarahnlewis commented on .gitignore:
```diff
@@ -132,3 +134,22 @@ cython_debug/
+uv.lock
+
+# samples
+*.pdb
```
This is problematic because there are .pdb files in the repo, referenced in tests. Instead of adding all this to .gitignore I suggest you don't save samples to the repo directory in the first place. Same comment for many of the other things added here.

sarahnlewis commented on src/bioemu/sample.py:
```diff
@@ -69,45 +75,119 @@ def main(
+            quadratically with the sequence length. A100-80GB would give you ~900 right at the memory
+            limit, so 500 is reasonable
+        model_name: Name of pretrained model to use. If this is set, you do not need to provide
+            `ckpt_path` or `model_config_path`. The model will be retrieved from huggingface; the
```
Is the biorxiv URL removed deliberately?

Is most of the diff in this docstring just new linebreaks?

sarahnlewis commented on src/bioemu/sample.py:
```diff
@@ -137,48 +217,99 @@ def main(
+        steering_config_dict = (
+            OmegaConf.create(steering_config_dict) if steering_config_dict is not None else None
+        )
+        print(
```
Remove this print statement or convert to logger.info and put it outside the loop over batches.

sarahnlewis commented on src/bioemu/sample.py:
```diff
@@ -188,9 +319,11 @@ def main(
     if set(sequences) != {sequence}:
         raise ValueError(f"Expected all sequences to be {sequence}, but got {set(sequences)}")
     positions = torch.tensor(np.concatenate([np.load(f)["pos"] for f in samples_files]))
+    # denoised_positions = torch.tensor(np.concatenate([np.load(f)["denoised_pos"] for f in samples_files], axis=0))
```
Remove commented code

sarahnlewis commented on src/bioemu/sample.py:
```diff
@@ -280,19 +421,24 @@ def generate_batch(
+    pos = torch.stack([x.pos for x in sampled_chemgraphs]).to("cpu")  # [BS, L, 3]
+    node_orientations = torch.stack([x.node_orientations for x in sampled_chemgraphs]).to(
+        "cpu"
+    )  # [BS, L, 3, 3]
```
thanks for adding the shape comments

sarahnlewis commented on src/bioemu/config/steering/physical_steering.yaml:
```diff
@@ -0,0 +1,48 @@
+  chainbreak:
+    # Enforces realistic Ca-Ca distances (3.8A) using flat-bottom loss
+    _target_: bioemu.steering.ChainBreakPotential
+    # flatbottom: Minimum allowed distance between non-neighboring Ca atoms (A) - zero penalty within this range
```
```suggestion
    # flatbottom: Maximum allowed distance between neighboring Ca atoms (A) - zero penalty within this range
```
?

Actually, I'd suggest to leave these comments out of this file and just document the parameters in the class definitions for ChainBreakPotential etc. so you only have to maintain docs in one place.

ludwigwinkler replied:
moved it to the class definition

sarahnlewis commented on src/bioemu/steering.py:
```diff
@@ -0,0 +1,333 @@
+    """
+    Compute R_0 given R_t and score.
+    """
+    alpha_t, sigma_t = sde.mean_coeff_and_std(x=R, t=t, batch_idx=batch_idx)
```
hm, is this legit when R lives in SO(3)? What do we need it for anyway? I thought the potentials were all defined in terms of CA positions.

Commit:
Refactor steering parameters and update configuration for consistency across codebase

Commit:
Add disulfide bridge steering potential and example scripts for comparison

ludwigwinkler commented on notebooks/disulfide_bridge_steering_example.py:
Quick numerical test to check disulfide bridge steering.

ludwigwinkler committed:
Merge branch 'main' into luwinkler/cli_steering

Commit:
Add type hints to log_physicality and potential_loss_fn functions for improved clarity

Commit:
Refactor logging in log_physicality function and add logger initialization

ludwigwinkler commented on pyproject.toml:
```diff
@@ -22,6 +22,9 @@ dependencies = [
     "typer",
     "uv",
+    "einops",
+    "matplotlib>=3.10.7",
```
the ppft notebook is using matplotib and my own steering example notebook does as well. Pandas is not required and I removed it.

Commit:
ReadMe update and typo fix

Commit:
reverted back to no return sample() function

Commit:
Remove sample result validation from steering tests for cleaner output

Commit:
Update steering section in README to clarify the use of steering particles

Commit:
black formatter pre-commit hook

Commit:
small ReadMe update

Commit:
Add integration tests for CLI functionality in BioEMU

This commit introduces a new test suite in `test_integration.py` that verifies the basic command from the README, steering functionality via CLI parameters, and the processing of steering parameters. The tests ensure that output files are generated correctly and that the integration works end-to-end.

sarahnlewis commented on src/bioemu/denoiser.py:
```diff
@@ -414,4 +432,67 @@ def dpm_solver(
+
+            total_energy = torch.stack(energies, dim=-1).sum(-1)  # [BS]
+
+            # if resampling implicitely given by num_fk_samples > 1
```
```suggestion
```

sarahnlewis commented on src/bioemu/denoiser.py:
```diff
@@ -414,4 +432,67 @@ def dpm_solver(
+
+                # ... or a single final sample
+                elif i >= N - 2:  # The last step is N-2
+                    print(
```
use logger

sarahnlewis commented on src/bioemu/denoiser.py:
```diff
@@ -1,16 +1,18 @@
-from .sde_lib import SDE, CosineVPSDE
-from .so3_sde import SO3SDE, apply_rotvec_to_rotmat
-
-TwoBatches = tuple[Batch, Batch]
```
ha good spot

ludwigwinkler commented:
@microsoft-github-policy-service agree company="Microsoft"

Commit:
Refactor disulfide bridge steering example to encapsulate main logic in a main guard

This commit restructures the `disulfide_bridge_steering_example.py` file by adding a main guard to encapsulate the sampling logic for structures with and without physicality steering. This improves code organization and allows for easier execution as a standalone script.

Commit:
Refactor denoiser and steering potentials to streamline input parameters and improve clarity

ludwigwinkler committed:
Merge branch 'main' into luwinkler/cli_steering

nw13slx reviewed (approved):
LGTM
